File: android/app/tests/unit/src/com/android/bluetooth/hfpclient/HeadsetClientServiceBinderTest.java
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.hfpclient;

import static android.bluetooth.BluetoothProfile.CONNECTION_POLICY_ALLOWED;
import static android.bluetooth.BluetoothProfile.STATE_CONNECTED;

import static com.android.bluetooth.TestUtils.getTestDevice;

import static org.mockito.Mockito.verify;

import android.bluetooth.BluetoothDevice;
import android.content.AttributionSource;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.SmallTest;

import com.android.tests.bluetooth.MockitoRule;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;

/** Test cases for {@link HeadsetClientServiceBinder}. */
@SmallTest
@RunWith(AndroidJUnit4.class)
public class HeadsetClientServiceBinderTest {
    @Rule public final MockitoRule mMockitoRule = new MockitoRule();

    @Mock private AttributionSource mSource;
    @Mock private HeadsetClientService mService;

    private final BluetoothDevice mDevice = getTestDevice(54);

    private HeadsetClientServiceBinder mBinder;

    @Before
    public void setUp() throws Exception {
        mBinder = new HeadsetClientServiceBinder(mService);
    }

    @Test
    public void connect_callsServiceMethod() {
        mBinder.connect(mDevice, mSource);

        verify(mService).connect(mDevice);
    }

    @Test
    public void disconnect_callsServiceMethod() {
        mBinder.disconnect(mDevice, mSource);

        verify(mService).disconnect(mDevice);
    }

    @Test
    public void getConnectedDevices_callsServiceMethod() {
        mBinder.getConnectedDevices(mSource);

        verify(mService).getConnectedDevices();
    }

    @Test
    public void getDevicesMatchingConnectionStates_callsServiceMethod() {
        int[] states = new int[] {STATE_CONNECTED};
        mBinder.getDevicesMatchingConnectionStates(states, mSource);

        verify(mService).getDevicesMatchingConnectionStates(states);
    }

    @Test
    public void getConnectionState_callsServiceMethod() {
        mBinder.getConnectionState(mDevice, mSource);

        verify(mService).getConnectionState(mDevice);
    }

    @Test
    public void setConnectionPolicy_callsServiceMethod() {
        int connectionPolicy = CONNECTION_POLICY_ALLOWED;
        mBinder.setConnectionPolicy(mDevice, connectionPolicy, mSource);

        verify(mService).setConnectionPolicy(mDevice, connectionPolicy);
    }

    @Test
    public void getConnectionPolicy_callsServiceMethod() {
        mBinder.getConnectionPolicy(mDevice, mSource);

        verify(mService).getConnectionPolicy(mDevice);
    }

    @Test
    public void startVoiceRecognition_callsServiceMethod() {
        mBinder.startVoiceRecognition(mDevice, mSource);

        verify(mService).startVoiceRecognition(mDevice);
    }

    @Test
    public void stopVoiceRecognition_callsServiceMethod() {
        mBinder.stopVoiceRecognition(mDevice, mSource);

        verify(mService).stopVoiceRecognition(mDevice);
    }

    @Test
    public void getAudioState_callsServiceMethod() {
        mBinder.getAudioState(mDevice, mSource);

        verify(mService).getAudioState(mDevice);
    }

    @Test
    public void setAudioRouteAllowed_callsServiceMethod() {
        boolean allowed = true;
        mBinder.setAudioRouteAllowed(mDevice, allowed, mSource);

        verify(mService).setAudioRouteAllowed(mDevice, allowed);
    }

    @Test
    public void getAudioRouteAllowed_callsServiceMethod() {
        mBinder.getAudioRouteAllowed(mDevice, mSource);

        verify(mService).getAudioRouteAllowed(mDevice);
    }

    @Test
    public void connectAudio_callsServiceMethod() {
        mBinder.connectAudio(mDevice, mSource);

        verify(mService).connectAudio(mDevice);
    }

    @Test
    public void disconnectAudio_callsServiceMethod() {
        mBinder.disconnectAudio(mDevice, mSource);

        verify(mService).disconnectAudio(mDevice);
    }

    @Test
    public void acceptCall_callsServiceMethod() {
        int flag = 2;
        mBinder.acceptCall(mDevice, flag, mSource);

        verify(mService).acceptCall(mDevice, flag);
    }

    @Test
    public void rejectCall_callsServiceMethod() {
        mBinder.rejectCall(mDevice, mSource);

        verify(mService).rejectCall(mDevice);
    }

    @Test
    public void holdCall_callsServiceMethod() {
        mBinder.holdCall(mDevice, mSource);

        verify(mService).holdCall(mDevice);
    }

    @Test
    public void terminateCall_callsServiceMethod() {
        mBinder.terminateCall(mDevice, null, mSource);

        verify(mService).terminateCall(mDevice, null);
    }

    @Test
    public void explicitCallTransfer_callsServiceMethod() {
        mBinder.explicitCallTransfer(mDevice, mSource);

        verify(mService).explicitCallTransfer(mDevice);
    }

    @Test
    public void enterPrivateMode_callsServiceMethod() {
        int index = 1;
        mBinder.enterPrivateMode(mDevice, index, mSource);

        verify(mService).enterPrivateMode(mDevice, index);
    }

    @Test
    public void dial_callsServiceMethod() {
        String number = "12532523";
        mBinder.dial(mDevice, number, mSource);

        verify(mService).dial(mDevice, number);
    }

    @Test
    public void sendDTMF_callsServiceMethod() {
        byte code = 21;
        mBinder.sendDTMF(mDevice, code, mSource);

        verify(mService).sendDTMF(mDevice, code);
    }

    @Test
    public void getCurrentAgEvents_callsServiceMethod() {
        mBinder.getCurrentAgEvents(mDevice, mSource);

        verify(mService).getCurrentAgEvents(mDevice);
    }

    @Test
    public void sendVendorAtCommand_callsServiceMethod() {
        int vendorId = 5;
        String cmd = "test_command";

        mBinder.sendVendorAtCommand(mDevice, vendorId, cmd, mSource);

        verify(mService).sendVendorAtCommand(mDevice, vendorId, cmd);
    }

    @Test
    public void getCurrentAgFeatures_callsServiceMethod() {
        mBinder.getCurrentAgFeatures(mDevice, mSource);

        verify(mService).getCurrentAgFeaturesBundle(mDevice);
    }

    @Test
    public void cleanUp_doesNotCrash() {
        mBinder.cleanup();
    }
}

File: android/app/tests/unit/src/com/android/bluetooth/hfpclient/HeadsetClientServiceBinderTest.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.hfpclient

import android.bluetooth.BluetoothProfile.CONNECTION_POLICY_ALLOWED
import android.bluetooth.BluetoothProfile.STATE_CONNECTED
import android.content.AttributionSource
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import com.android.bluetooth.getTestDevice
import com.android.tests.bluetooth.MockitoRule
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.kotlin.verify

/** Test cases for [HeadsetClientServiceBinder]. */
@SmallTest
@RunWith(AndroidJUnit4::class)
class HeadsetClientServiceBinderTest {
    @get:Rule val mockitoRule = MockitoRule()

    @Mock private lateinit var source: AttributionSource
    @Mock private lateinit var service: HeadsetClientService

    private val device = getTestDevice(54)

    private lateinit var binder: HeadsetClientServiceBinder

    @Before
    fun setUp() {
        binder = HeadsetClientServiceBinder(service)
    }

    @Test
    fun connect_callsServiceMethod() {
        binder.connect(device, source)

        verify(service).connect(device)
    }

    @Test
    fun disconnect_callsServiceMethod() {
        binder.disconnect(device, source)

        verify(service).disconnect(device)
    }

    @Test
    fun getConnectedDevices_callsServiceMethod() {
        binder.getConnectedDevices(source)

        verify(service).connectedDevices
    }

    @Test
    fun getDevicesMatchingConnectionStates_callsServiceMethod() {
        val states = intArrayOf(STATE_CONNECTED)
        binder.getDevicesMatchingConnectionStates(states, source)

        verify(service).getDevicesMatchingConnectionStates(states)
    }

    @Test
    fun getConnectionState_callsServiceMethod() {
        binder.getConnectionState(device, source)

        verify(service).getConnectionState(device)
    }

    @Test
    fun setConnectionPolicy_callsServiceMethod() {
        val connectionPolicy = CONNECTION_POLICY_ALLOWED
        binder.setConnectionPolicy(device, connectionPolicy, source)

        verify(service).setConnectionPolicy(device, connectionPolicy)
    }

    @Test
    fun getConnectionPolicy_callsServiceMethod() {
        binder.getConnectionPolicy(device, source)

        verify(service).getConnectionPolicy(device)
    }

    @Test
    fun startVoiceRecognition_callsServiceMethod() {
        binder.startVoiceRecognition(device, source)

        verify(service).startVoiceRecognition(device)
    }

    @Test
    fun stopVoiceRecognition_callsServiceMethod() {
        binder.stopVoiceRecognition(device, source)

        verify(service).stopVoiceRecognition(device)
    }

    @Test
    fun getAudioState_callsServiceMethod() {
        binder.getAudioState(device, source)

        verify(service).getAudioState(device)
    }

    @Test
    fun setAudioRouteAllowed_callsServiceMethod() {
        val allowed = true
        binder.setAudioRouteAllowed(device, allowed, source)

        verify(service).setAudioRouteAllowed(device, allowed)
    }

    @Test
    fun getAudioRouteAllowed_callsServiceMethod() {
        binder.getAudioRouteAllowed(device, source)

        verify(service).getAudioRouteAllowed(device)
    }

    @Test
    fun connectAudio_callsServiceMethod() {
        binder.connectAudio(device, source)

        verify(service).connectAudio(device)
    }

    @Test
    fun disconnectAudio_callsServiceMethod() {
        binder.disconnectAudio(device, source)

        verify(service).disconnectAudio(device)
    }

    @Test
    fun acceptCall_callsServiceMethod() {
        val flag = 2
        binder.acceptCall(device, flag, source)

        verify(service).acceptCall(device, flag)
    }

    @Test
    fun rejectCall_callsServiceMethod() {
        binder.rejectCall(device, source)

        verify(service).rejectCall(device)
    }

    @Test
    fun holdCall_callsServiceMethod() {
        binder.holdCall(device, source)

        verify(service).holdCall(device)
    }

    @Test
    fun terminateCall_callsServiceMethod() {
        binder.terminateCall(device, null, source)

        verify(service).terminateCall(device, null)
    }

    @Test
    fun explicitCallTransfer_callsServiceMethod() {
        binder.explicitCallTransfer(device, source)

        verify(service).explicitCallTransfer(device)
    }

    @Test
    fun enterPrivateMode_callsServiceMethod() {
        val index = 1
        binder.enterPrivateMode(device, index, source)

        verify(service).enterPrivateMode(device, index)
    }

    @Test
    fun dial_callsServiceMethod() {
        val number = "12532523"
        binder.dial(device, number, source)

        verify(service).dial(device, number)
    }

    @Test
    fun sendDTMF_callsServiceMethod() {
        val code: Byte = 21
        binder.sendDTMF(device, code, source)

        verify(service).sendDTMF(device, code)
    }

    @Test
    fun getCurrentAgEvents_callsServiceMethod() {
        binder.getCurrentAgEvents(device, source)

        verify(service).getCurrentAgEvents(device)
    }

    @Test
    fun sendVendorAtCommand_callsServiceMethod() {
        val vendorId = 5
        val cmd = "test_command"

        binder.sendVendorAtCommand(device, vendorId, cmd, source)

        verify(service).sendVendorAtCommand(device, vendorId, cmd)
    }

    @Test
    fun getCurrentAgFeatures_callsServiceMethod() {
        binder.getCurrentAgFeatures(device, source)

        verify(service).getCurrentAgFeaturesBundle(device)
    }

    @Test
    fun cleanUp_doesNotCrash() {
        binder.cleanup()
    }
}
